### PR TITLE
created enumerable methods

### PR DESCRIPTION
**Pull request by gitnyasha**
A module containing enumerable methods for use

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    return to_module unless block_given?
+
+    x = 0
+    while x < length
+      yield(to_a[x])
+      x += 1
+    end
+    self
+  end
+
```
Comment (imhta) [inline enumerable.rb]: * [ ] should have working `my_each`  identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### Errors
```bash
rspec ./spec/spec_enumerables_spec.rb:27 # enumerables my_each returns an Enumerator if no block is given
```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  def my_each_with_index
+    return to_module unless block_given?
+
+    x = 0
+    while x < length
+      yield(self[x], x)
+      x += 1
+    end
+    self
+  end
```
Comment (imhta) [inline enumerable.rb]: * [ ] should have working `my_each_with_index`  identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Errors
```bash
rspec ./spec/spec_enumerables_spec.rb:43 # enumerables my_each_with_index returns an enumerator if no block is given
```

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    return to_module unless block_given?
+
+    new_arr = []
+    if block_given?
+      my_each { |x| new_arr << yield(x) }
+    else
+      my_each { |x| new_arr << param.call(x) }
+    end
+    new_arr
+  end
```
Comment (imhta) [inline enumerable.rb]: * [ ] should have working `my_map`  identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Errors
```bash
rspec ./spec/spec_enumerables_spec.rb:200 # enumerables my_map returns an Enumerator if no block is given
```

**Review by imhta — CHANGES_REQUESTED** [review]
# Changes required 🛑 

### Hello @gitnyasha  , 👋 There some tests are failing in your methods. kindly check the inline the comments, and fix them.

* [x] should have working `my_select` **(covered by automated tests)** identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Errors
```bash
rspec ./spec/spec_enumerables_spec.rb:49 # enumerables my_select returns an array containing all elements of enum for which the given block returns a true value
rspec ./spec/spec_enumerables_spec.rb:53 # enumerables my_select returns an enumerator if no block is given
```
----
#### Don't forgot to request for another code review
----

**Pon Muthu Selvam N**
_If you have any queries feel free to ping me on slack_

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice progress :+1: 
I see you have come a long way with this project and we also have a lot to do before we can merge this one. See my comments for suggestions.

### Martin Afani

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      i = 0
+      while i < results.length
+        yield(results[i])
+        i += 1
+      end
+      results
+    else
+      results.to_enum
+    end
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] calls the given block once for each element in self
- [ ] returns an Enumerator if no block is given

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      i = 0
+      while i < results.length
+        yield(results[i], i)
+        i += 1
+      end
+      results
+    else
+      results.to_enum
+    end
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] calls the given block once for each element in self
- [ ] returns an enumerator if no block is given

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+      results.my_each do |x|
+        if yield(x)
+          choice.push(x)
+        end
+      end
+      choice
+    else
+      results.to_enum
+    end
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns an array containing all elements of enum for which the given block returns a true value
- [ ] returns an enumerator if no block is given

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      if items.class == Regexp
+        my_each { |x| return false unless items.match?(x) }
+      elsif items.class == Class
+        my_each { |x| return false unless x.is_a? items }
+      else
+        my_each { |x| return false unless x == items }
+      end
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns true if the block never returns false or nil
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    result = false
+    if block_given?
+      my_each { |item| result = true if yield item }
+    elsif items
+      my_each { |item| result = true if itemstern?(item, items) }
+    else
+      my_each { |item| result = true if item }
+    end
+    result
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns true if the block ever returns a value other than false or nil
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    result = true
+    if block_given?
+      my_each { |item| result = false if yield item }
+    elsif items
+      my_each { |item| result = false if itemstern?(item, items) }
+    else
+      my_each { |item| result = false if item }
+    end
+    result
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns true if the block never returns true for all elements
- [ ] when no block or argument is given returns true only if none of the collection members is true
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    count = 0
+    if block_given?
+      my_each { |x| count += 1 if yield(x) }
+    elsif arg
+      my_each { |x| count += 1 if x == arg }
+    else
+      count = length
+    end
+    count
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] counts the number of items in enum that are equal to item if an argument is given
- [ ] counts the number of elements yielding a true value if a block is given

File: enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    if block_given?
+      numbers = []
+      results.my_each do |x|
+        numbers << yield(x)
+      end
+      numbers
+    else
+      results.to_enum
+    end
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns a new array with the results of running block once for every element in enum.
- [ ] returns an Enumerator if no block is given

File: enumerable.rb [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    result, sample = inj_param(*args)
+    arr = result ? to_a : to_a[1..-1]
+    result ||= to_a[0]
+    if block_given?
+      arr.my_each { |x| result = yield(result, x) }
+    elsif sample
+      arr.my_each { |x| result = result.public_send(sample, x) }
+    end
+    result
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] combines all elements of enum by applying a binary operation, specified by a block
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method

Comment (gitnyasha) [reply enumerable.rb]: can you give an example of one of the methods

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Hey @gitnyasha 
I know you are trying your level best to meet the requirements of this project which is good. But we have a long way to go. Your methods fail against the same edge cases as before. Please take your time to fix the failing cases.
#### Look at my suggestions from the previous review.

### Martin Afani

File: enumerable.rb (hunk 12 above)
Comment (gitnyasha) [reply enumerable.rb]: Can you give an example of what I need to do because I dont know what you want me to do

Comment (whiz25) [reply enumerable.rb]: The whole idea is for your methods to behave the same way as the original [enumerable methods](https://ruby-doc.org/core-2.7.0/Enumerable.html).
For example, 
```ruby
let(:range) { Range.new(5, 50) }
actual = range.my_inject(4) { |prod, n| prod * n }
expected = range.inject(4) { |prod, n| prod * n }
expect(actual).to eq expected
```
Should return true if your method `my_inject` works as expected.

Comment (gitnyasha) [reply enumerable.rb]: Ok

File: enumerable.rb [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_select
+    item = []
+    self.my_each do |i|
+      item << i if yield(i)
+    end
+    item
+  end
+
+  def my_all?
```
Comment (Forison) [inline enumerable.rb:26]: * [ ] Run your methods against these examples,
```  p %w[ant bear cat].my_all? { |word| word.length >= 3 } #=> true
     p %w[ant bear cat].my_all? { |word| word.length >= 4 } #=> false
     p %w[ant bear cat].my_all?(/t/)                        #=> false
     p [1, 2i, 3.14].my_all?(Numeric)                       #=> true
     p [nil, true, 99].my_all?                              #=> false
     p [].my_all?                                           #=> true
```
compare the result with
```  p %w[ant bear cat].all? { |word| word.length >= 3 } #=> true
     p %w[ant bear cat].all? { |word| word.length >= 4 } #=> false
     p %w[ant bear cat].all?(/t/)                        #=> false
     p [1, 2i, 3.14].all?(Numeric)                       #=> true
     p [nil, true, 99].all?                              #=> false
     p [].all?                                           #=> true```
both ```.all``` and ```.my_all``` should be same

File: enumerable.rb [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    item = []
+    self.my_each do |i|
+      item << i if yield(i)
+    end
+
+    return true if item.size == self.size
+    return
+  end
+
+  def my_any?
```
Comment (Forison) [inline enumerable.rb:32]: * [ ] compare results from the custom method with the original
```
 p %w[ant bear cat].my_any?(/d/)                        #=> false
  p [nil, true, 99].my_any?(Integer)                     #=> true
  p [nil, true, 99].my_any?                              #=> true
  p [].my_any?
```
original method
```
  p %w[ant bear cat].any?(/d/)                        #=> false
  p [nil, true, 99].any?(Integer)                     #=> true
  p [nil, true, 99].any?                              #=> true
  p [].any?
```

File: enumerable.rb [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    item = []
+    self.my_each do |i|
+      item << i if yield(i)
+    end
+
+    return true if item.size > 0
+    return
+  end
+
+  def my_none?
```
Comment (Forison) [inline enumerable.rb:38]: * [ ] Do same for this ```.my_none```, compare reutls with ```.none```
```p %w{ant bear cat}.my_none? { |word| word.length == 5 } #=> true
p %w{ant bear cat}.my_none? { |word| word.length >= 4 } #=> false
p %w{ant bear cat}.my_none?(/d/)                        #=> true
p [1, 3.14, 42].my_none?(Float)                         #=> false
p [].my_none?                                           #=> true
p [nil].my_none?                                        #=> true
p [nil, false].my_none?                                 #=> true
p [nil, false, true].my_none?                           #=> false```

File: enumerable.rb [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    arr = []
+
+    self.my_each do |i|
+      arr << yield(i)
+    end
+
+    arr
+  end
+
+  def my_inject
```
Comment (Forison) [inline enumerable.rb]: * [ ] Do same for ```my_inject```, compare results of example with results from original ```inject```.
```p (5..10).my_inject { |sum, n| sum + n }
p (5..10).my_inject(1) { |product, n| product * n } #=> 151200
longest = %w{ cat sheep bear }.my_inject do |memo, word|
   memo.length > word.length ? memo : word
end
p longest ```

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:Require_changes🛑
# Hi @gitnyasha🙌,
I see you have come a long way so far😇. Don't worry coding is about problem-solving😇. Eventually, you learn a lot😇.
Like previous reviewers said, for this project, you are to create custom methods that have the same functionalities as their original version. For each of the methods, get info about their functionality from the [ruby enum doc](https://ruby-doc.org/core/Enumerable.html) and replicate it🙂.
> visit [ruby enum docs](https://ruby-doc.org/core/Enumerable.html) for functionalities and examples to test your method.
## Check below for how we can resolve the errors you are having👇
### Kindly make these changes and submit the project for review.
#### Happy coding
#### Addo Forison

File: enumerable.rb (hunk 16 above)
Comment (gitnyasha) [reply enumerable.rb]: Thank you

**Review by whiz25 — APPROVED** [review]
## Hey @gitnyasha 
You have tried your level best with this project. Clearly you have a long way to go but because we have exceeded the maximum number of review cycles for this project, I am approving it. Please merge and proceed.

### Martin Afani

**Pull request merged**